### PR TITLE
prototype(card): create prototype card based on MDC web

### PR DESCRIPTION
**Pull request by jelbourn**

### CODE REVIEW

File: src/material-experimental/mdc-card/BUILD.bazel [hunk 1]
```diff
@@ -22,13 +22,22 @@ sass_library(
...
 sass_binary(
     name = "card_scss",
     src = "card.scss",
+    include_paths = [
+        "external/npm/node_modules",
+    ],
+    deps = [
+        "//src/material-experimental/mdc-helpers:mdc_helpers_scss_lib",
+        "//src/material-experimental/mdc-helpers:mdc_scss_deps_lib",
+        "//src/material/core:all_themes",
```
Comment (devversion) [inline src/material-experimental/mdc-card/BUILD.bazel]: why do we bring in the whole theme here? shouldn't we just fine-grain what is used?

This generally causes unnecessary rebuilds if unrelated things change and the Sass compilation is not very fast right now (it's a bottleneck)

File: src/material-experimental/mdc-card/card.scss [hunk 2]
```diff
@@ -1 +1,169 @@
...
+
+  // Apply default padding for the header region. Omit any bottom padding because we assume
+  // this region will be followed by another region that includes top padding.
+  padding: $mat-card-default-padding $mat-card-default-padding 0;
+
+  // When a subtitle is inside of a header, we want to move it up slightly to reduce the space with
+  // the title, and add a margin bottom to create space underneath the header.
+  .mat-mdc-card-subtitle {
+    margin-top: -8px;
+    margin-bottom: 16px;
```
Comment (devversion) [inline src/material-experimental/mdc-card/card.scss]: Just asking out of curiosity: can't we calculate this somehow through variables? (it feels a bit like magic numbers to me)

File: src/material-experimental/mdc-card/card.scss [hunk 3]
```diff
@@ -1 +1,169 @@
...
+
+// When both a title and a subtitle are used, eliminate the top padding of whichever comes second
+// because the first already covers the padding.
+.mat-mdc-card-subtitle ~ .mat-mdc-card-title,
+.mat-mdc-card-title ~ .mat-mdc-card-subtitle {
+  padding-top: 0;
+}
+
+// When the card actions come after the footer, eliminate the action's bottom padding because
+// it's caputred by the footer.
```
Comment (devversion) [inline src/material-experimental/mdc-card/card.scss]:
```suggestion
// it's captured by the footer.
```

Comment (devversion) [reply src/material-experimental/mdc-card/card.scss]: Also can you explain what is happening here? I'm confused by the comment that says that the footer captures the bottom padding of the actions? how can that work if the footer is an element before the actions?


 If I understand correctly: if actions come after the footer, the actions don't need a bottom padding?

File: src/material-experimental/mdc-card/BUILD.bazel (hunk 1 above)
Comment (jelbourn) [reply src/material-experimental/mdc-card/BUILD.bazel]: Good call, I just copy-pasted from the checkbox without thinking about it. It's actually not necessary at all.

File: src/material-experimental/mdc-card/card.scss (hunk 2 above)
Comment (jelbourn) [reply src/material-experimental/mdc-card/card.scss]: Changed to use `$mat-card-default-padding`

File: src/material-experimental/mdc-card/card.scss (hunk 3 above)
Comment (jelbourn) [reply src/material-experimental/mdc-card/card.scss]: You're right, this makes no sense. Removed.

File: src/material-experimental/mdc-card/card.scss [hunk 4]
```diff
@@ -1 +1,163 @@
...
+// Avatar image for a card. MDC does not specifically have a card avatar or a "common" avatar.
+// They *do* have a list avatar, but it uses a different size than we do here, which we preserve
+// to reduce breaking changes. Ultimately, the avatar styles just consist of a size and a
+// border-radius.
+.mat-mdc-card-avatar {
+  height: $mat-card-header-size;
+  width: $mat-card-header-size;
+  border-radius: 50%;
+  flex-shrink: 0;
+
```
Comment (crisbeto) [inline src/material-experimental/mdc-card/card.scss]: Does this also need a `background-position: center`.

File: src/material-experimental/mdc-card/card.scss [hunk 5]
```diff
@@ -1 +1,163 @@
...
+  .mat-mdc-card-title,
+  .mat-mdc-card-subtitle {
+    padding-top: 0;
+  }
+}
+
+// When both a title and a subtitle are used, eliminate the top padding of whichever comes second
+// because the first already covers the padding.
+.mat-mdc-card-subtitle ~ .mat-mdc-card-title,
+.mat-mdc-card-title ~ .mat-mdc-card-subtitle {
```
Comment (crisbeto) [inline src/material-experimental/mdc-card/card.scss]: You could potentially combine it with the declaration above since it's declaring the same styles.

File: src/material-experimental/mdc-helpers/_mdc-helpers.scss [hunk 6]
```diff
@@ -114,7 +114,7 @@ $mat-typography-level-mappings: (
   $primary: mat-color(map-get($theme, primary));
   $accent: mat-color(map-get($theme, accent));
   $warn: mat-color(map-get($theme, warn));
-  $background: mat-color(map-get($theme, background), background);
+  $backgroundPalette: map-get($theme, background);
```
Comment (crisbeto) [inline src/material-experimental/mdc-helpers/_mdc-helpers.scss]: Shouldn't this be dash-cased?

File: src/material-experimental/mdc-card/module.ts [hunk 7]
```diff
@@ -9,12 +9,45 @@
...
+  MatCardTitle,
+  MatCardTitleGroup,
+  MatCardXlImage
+];
 
 @NgModule({
   imports: [MatCommonModule, CommonModule],
-  exports: [MatCard, MatCommonModule],
-  declarations: [MatCard],
+  exports: CARD_DIRECTIVES,
```
Comment (mmalerba) [inline src/material-experimental/mdc-card/module.ts]: Do we still need to export `MatCommonModule`?

File: src/material-experimental/mdc-card/card.scss [hunk 8]
```diff
@@ -1 +1,163 @@
...
+.mat-mdc-card-content {
+  // Custom elements default to `display: inline`. Reset to 'block'.
+  display: block;
+
+  // Apply default horizontal padding for a content region.
+  padding: 0 $mat-card-default-padding;
+
+  // Only add vertical padding to the main content area if it's not preceeded/followed by another
+  // element within the card.
+  &:first-child {
```
Comment (jelbourn) [inline src/material-experimental/mdc-card/card.scss]: Look at using first-of-type with a common section class

File: src/material-experimental/mdc-card/card.scss (hunk 5 above)
Comment (jelbourn) [reply src/material-experimental/mdc-card/card.scss]: Done, though it does make the source less pretty since you have to expand the nesting.
